org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/codegen/BranchLabel.java

```diff
@@ -116,7 +117,39 @@ public void becomeDelegateFor(BranchLabel otherLabel) {
 	this.forwardReferences = mergedForwardReferences;
 	this.forwardReferenceCount = indexInMerge;
 }
-
+private boolean shouldThrow = true;
+protected void trackStackDepth(boolean branch) {
+	/* Control can reach an instruction with a label in two ways: (1) via a branch using that label or (2) by falling through from the previous instruction.
+	   In both cases, we KNOW the stack depth at the instruction from which control flows to the instruction with the label.
+	   Control cannot reach the instruction with a label by falling through if the previous instruction completes abruptly via a goto/return/throw etc
+	   this.codeStream.lastAbruptCompletion == this.position ==> instruction prior to the one with the label is a goto/return/throw
+	*/
+	boolean sourceDepthKnown = branch || this.codeStream.lastAbruptCompletion != this.position;
```

Comment:
I think we should be resetting the fall-throught stack depth to zero
ie the code can modified as the one shown below:

`	if (this.codeStream.lastAbruptCompletion == this.position) { // fall-through
		this.codeStream.stackDepth = this.targetStackDepth > 0 ? this.targetStackDepth : 0;
		return;
	}
	boolean sourceDepthKnown = branch; // can just use branch now instead of sourceDepthKnown
`

Reply:
Let us discuss this case:

```
public class X {
	static void foo(int a, int b, int c) {
		System.out.println(c);
	}
	public static void main(String[] args) {
		args = new String [] { "Hello", "World" };
		
		foo(10, 20, switch(args.length) {
			default -> {
				int i = args.length;
				int j = 0;
				while (i > 0) {
					i--;
					j++;
				}
				yield j;
			}
		});
	}
}
```

which results in this bytecode:

```
  public static void main(java.lang.String[]);
    descriptor: ([Ljava/lang/String;)V
    flags: (0x0009) ACC_PUBLIC, ACC_STATIC
    Code:
      stack=4, locals=3, args_size=1
         0: iconst_2
         1: anewarray     #34                 // class java/lang/String
         4: dup
         5: iconst_0
         6: ldc           #36                 // String Hello
         8: aastore
         9: dup
        10: iconst_1
        11: ldc           #38                 // String World
        13: aastore
        14: astore_0
        15: bipush        10
        17: bipush        20
        19: aload_0
        20: arraylength
        21: pop
        22: aload_0
        23: arraylength
        24: istore_1
        25: iconst_0
        26: istore_2
        27: goto          36
        30: iinc          1, -1
        33: iinc          2, 1
        36: iload_1
        37: ifgt          30
        40: iload_2
        41: invokestatic  #40                 // Method foo:(III)V
        44: return
```

At bci 30, we come in with a `stackDepth` of 2 - this is the correct value - we should not reset it to 0.

Reply:
> I think we should be resetting the stack depth of the instruction following the abrupt completion/goto to zero ie the code can modified as the one shown below:
> 
> ```
>         if (this.codeStream.lastAbruptCompletion == this.position) { 
> 
> 		this.codeStream.stackDepth = this.targetStackDepth > 0 ? this.targetStackDepth : 0;
> 
> 		return;
> 
> 	}
> 	boolean sourceDepthKnown = branch; // can just use branch now instead of sourceDepthKnown
> ```

What about the case `!branch && this.codeStream.lastAbruptCompletion != this.position` ??